server/sms/sms_gateway.go

```diff
@@ -188,43 +160,34 @@ func (g *gateway) proxyLoop() (*protocol.Message, error) {
 			}
 
 			err := g.send(receivedMsg)
-			if err != nil {
-				return receivedMsg, err
+			if err != nil && err == ErrRetryFailed {
+				// THIS MAY BE BLOCKING.Maybe not a good idea.
+				for err2 := g.SetLastSentID(receivedMsg.ID); err2 != nil; {
+					g.logger.WithField("error", err2.Error()).Error("Error setting last ID.Retrying")
+					time.Sleep(100 * time.Millisecond)
+				}
+				g.logger.WithField("id", receivedMsg.ID).Info("Set last id to ")
+				continue
+			} else if err != nil {
+				g.logger.WithField("err", err.Error()).Error("Exiting from proxyLoop.")
+				return err
+			} else {
+				continue
 			}
+
 		case <-g.ctx.Done():
 			// If the parent context is still running then only this subscriber context
 			// has been cancelled
 			if g.ctx.Err() == nil {
-				return nil, g.ctx.Err()
+				return g.ctx.Err()
```

Comment:
But be careful and error can appear in case of canceling the context, which happens on `Stop` method